XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+
+List of all providers that may trigger the `init` callback.
+
+### Constructor: `AsyncHook`
+
+The `AsyncHook` constructor returns an instance that contains information about
+the callbacks that are to fire during specific asynchronous events in the
+lifetime of the event loop. The focal point of these calls centers around the
+lifetime of `AsyncWrap`. These callbacks will also be called to emulate the
+lifetime of handles and requests that do not fit this model. For example,
+`HTTPPARSER` instances are recycled to improve performance. So the `destroy()`
+callback would be called manually after a connection is done using it, just
+before it's placed back into the unused resource pool.
+
+All callbacks are optional. So if only resource cleanup needs to be tracked
+then only the `destroy()` callback needs to be passed.
+
+**Error Handling**: If any callback throws the application will print the
+stack trace and exit. The exit path does follow that of any uncaught exception,
+so any `'exit'` callbacks will fire. Unless the application is run with
+`--abort-on-uncaught-exception`. In which case a stack trace will be printed
+and the application will exit, leaving a core file.
+
+The reason for this behavior is that these callbacks are running at potentially
+volatile points in an object's lifetime. For example during class construction
+and destruction. Because of this, it is deemed necessary to bring down the
+process quickly as to prevent an unintentional abort in the future. This is
+subject to change in the future if a comprehensive analysis is performed to
+ensure an exception can follow the normal control flow without unintentional
+side effects.
+
+#### `asyncHook.enable()`
+
+Enable the callbacks for a given `AsyncHook` instance. Once a callback fires on
+an asynchronous event they will continue to fire for all nested asynchronous
+events. Even after the instance has been disabled.
+
+Callbacks are not implicitly enabled after an instance is created. Reason for
+this is to not make any assumptions about the user's use case. Since
+constructing the `asyncHook` during startup but not using it until later is a
+perfectly reasonable use case. This API is meant to err on the side of
+requiring explicit instructions from the user.
+
+#### `asyncHook.disable()`
+
+Disable the callbacks for a given `AsyncHook` instance. Doing this will prevent
+the `init()`, etc., calls from firing for any new roots of asynchronous call
+stacks, but will not prevent existing asynchronous call stacks that have
+already been captured by the `AsyncHook` instance from continuing to fire.
+
+While not part of the immediate development plan, it should be possible in the
+future to allow selective tracking of asynchronous call stacks. The following
+example demonstrates this:
+
+```js
+const async_wrap = require('async_wrap');
+const net = require('net');
+
+// Pretend init, pre, post are all defined
+const asyncHook = async_wrap.createHook({ init, pre, post });
```

Comment:
As far as I know, almost entirely dependant on the work done from the hook than the hook itself.